### PR TITLE
Tidied up Ethos-U delegate binaries 

### LINKED ISSUE #677: Removed std::vector container from the Arm delegate runtime code
Let's try to see if we can get away without this std::container. It will be better from size and portability point of view. Other than input/output we aren't using it anywhere else so should be easy to get rid of it.

_Originally posted by @digantdesai in https://github.com/pytorch/executorch/pull/659#discussion_r1349434360_

### PR DESCRIPTION
**Pull request by robell**
fixed #677 
simplified and made binary format more robust/validate-able and moved more processing to AoT
made runtime input/output support more general (arbitrary number of EValue inputs and outputs)

### CODE REVIEW

File: backends/arm/arm_backend.py [hunk 1]
```diff
@@ -145,8 +145,9 @@ def dbg_tosa_dump(tosa_fb, path):
         f.write(js)
 
 
-# Output to Vela with current file-based compilation
-# WARNING: if this changes, the runtime reader also needs to change
+# Output via Vela to binary stream for ArmBackendEthosU
+# WARNING: Do not change this without changing VelaBinStream.cpp as that
```
Comment (digantdesai) [inline backends/arm/arm_backend.py:149]: once we figure out e2e testing, this should be tested

File: backends/arm/arm_backend.py [hunk 2]
```diff
@@ -162,65 +163,82 @@ def vela_compile(tosa_fb):
...
+            # copy command data through unmodified
+            bin_blocks["cmd_data"] = data["cmd_data"].tobytes()
+
+            # copy weight data through unmodified
+            bin_blocks["weight_data"] = data["weight_data"].tobytes()
+
+            # Add a block for scratch, inputs and outputs;  scratch shape is a 1 element
+            # array giving us size in bytes so extract this and add a block of 0's.
+            # Currently we preallocated this on the host to provide SRAM for computation.
+            block_length = data["scratch_shape"][0].item()
```
Comment (digantdesai) [inline backends/arm/arm_backend.py:181]: nit: assert length of 1

File: backends/arm/arm_backend.py [hunk 3]
```diff
@@ -162,65 +163,82 @@ def vela_compile(tosa_fb):
...
+            for i in range(len(data["input_shape"])):
+                input_shape = data["input_shape"][i]
+                input_elem_size = data["input_elem_size"][i]
+                input_offset = data["input_offset"][i]
+                input_region = data["input_region"][i]
+                assert len(input_shape) <= 4
+                inp_pad = input_shape.tolist() + [0] * (4 - len(input_shape))
+                input_struct = struct.pack(
+                    "<iiiiiii", *inp_pad, input_elem_size, input_offset, input_region
+                )
```
Comment (digantdesai) [inline backends/arm/arm_backend.py:196]: make a function?
`def create_struct(data: Dict[str, Any], type: Enum, index: int) -> Struct:`
type can be an enum of input, output

File: backends/arm/runtime/VelaBinStream.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      if (ptr - data != size)
+        return false;
+      return true;
+    } else {
+      // Unrecognised block name
+      return false;
+    }
+  }
+
+  // We've fallen off the end without finding vela_end_stream
```
Comment (digantdesai) [inline backends/arm/runtime/VelaBinStream.cpp:78]: same here

File: backends/arm/runtime/VelaBinStream.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+#pragma once
+
+#include <cstdint>
+
+// Generic block within the vela_bin_stream encoded by the python vela_compile
+// step
+typedef struct {
+  char name[16]; // string name, can be shorter or truncated
```
Comment (digantdesai) [inline backends/arm/runtime/VelaBinStream.h:23]:
```suggestion
  char name[kVelaBlockNameLength]; // string name, can be shorter or truncated
```

File: backends/arm/runtime/VelaBinStream.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  size_t cmd_data_size;
+  const char* weight_data;
+  size_t weight_data_size;
+  const char* scratch_data;
+  size_t scratch_data_size;
+  VelaIOs* inputs;
+  VelaIOs* outputs;
+} VelaHandles;
+
+bool vela_bin_read(const char* data, VelaHandles* handles, int size);
```
Comment (digantdesai) [inline backends/arm/runtime/VelaBinStream.h:55]: add a comment for this API and one below?

File: examples/arm/aot_arm_compiler.py [hunk 7]
```diff
@@ -38,6 +38,34 @@ def forward(self, x):
...
+        return x + y
+
+    example_input = (
+        torch.ones(5, dtype=torch.int32),
+        torch.ones(5, dtype=torch.int32),
+    )
+    can_delegate = True
+
+
+class AddModule3(torch.nn.Module):
```
Comment (digantdesai) [inline examples/arm/aot_arm_compiler.py:55]: can we add something with weights? I was trying Linear and it was failing with an error in vela

File: examples/arm/run.sh [hunk 8]
```diff
@@ -34,23 +34,20 @@ fvp_model=FVP_Corstone_SSE-300_Ethos-U55
...
-    pte_file="$(realpath ./softmax.pte)"
-    [[ -f ${pte_file} ]] || { echo "Failed to generate a pte file - ${pte_file}"; exit 1; }
-    echo "${pte_file}"
-}
-
-# Generate the ethos delegate PTE file
-function generate_ethos_pte_file() {
-    cd $et_root_dir
-    python3 -m examples.arm.aot_arm_compiler --model_name="add" --delegate 1>&2
+    python3 -m examples.arm.aot_arm_compiler --model_name="${model}" ${delegate} 1>&2
```
Comment (digantdesai) [inline examples/arm/run.sh:48]: nice!

File: examples/arm/run.sh [hunk 9]
```diff
@@ -150,16 +147,17 @@ type ${buck2} 2>&1 > /dev/null \
...
-pte_delegate=$(generate_ethos_pte_file)
-# build and run the same app with a delegated .pte
-build_executorch_runner "${pte_delegate}"
-run_fvp executor_runner.elf
+# the test models run, and whether to delegate
+test_model=( "softmax" "add" "add3" )
+test_delegate=( "" "--delegate" "--delegate" )
+
+# loop over running the AoT flow and executing the model on device
+for i in "${!test_model[@]}"; do
```
Comment (digantdesai) [inline examples/arm/run.sh:155]: reaching the complexity limit of bash LOL

**Review by digantdesai — APPROVED** [review]
LGTM. Added some nit comments, let me know when you are ready and I can merge it for you :)

File: backends/arm/arm_backend.py (hunk 1 above)
Comment (robell) [reply backends/arm/arm_backend.py:149]: I've structured the VelaBin* as having no dependencies and i have a local c test harness to enable a quick test of the serialisation form, would aim to structure it into unit test of serialisation.

File: backends/arm/arm_backend.py (hunk 2 above)
Comment (robell) [reply backends/arm/arm_backend.py:201]: fixed.

File: backends/arm/arm_backend.py (hunk 3 above)
Comment (robell) [reply backends/arm/arm_backend.py]: fixed.

File: backends/arm/runtime/VelaBinStream.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    } else if (!strncmp(b->name, "outputs", strlen("outputs"))) {
+      handles->outputs = (VelaIOs*)b->data;
+    } else if (!strncmp(
+                   b->name, "vela_end_stream", strlen("vela_end_stream"))) {
+      // expect vela_end_stream last
+      if (ptr - data != size)
+        return false;
+      return true;
+    } else {
+      // Unrecognised block name
```
[the remark this answers is not in the document]
Comment (robell) [reply backends/arm/runtime/VelaBinStream.cpp:77]: fixed.

File: backends/arm/runtime/VelaBinStream.h (hunk 5 above)
Comment (robell) [reply backends/arm/runtime/VelaBinStream.h]: fixed.

File: backends/arm/runtime/VelaBinStream.h (hunk 6 above)
Comment (robell) [reply backends/arm/runtime/VelaBinStream.h:62]: fixed.

File: examples/arm/aot_arm_compiler.py (hunk 7 above)
Comment (robell) [reply examples/arm/aot_arm_compiler.py:55]: It's on the list, but i'd aim to do this separately in an "improve vela" set of changes.

File: backends/arm/runtime/VelaBinStream.cpp (hunk 4 above)
Comment (robell) [reply backends/arm/runtime/VelaBinStream.cpp:83]: fixed

File: backends/arm/arm_backend.py (hunk 1 above)
Comment (digantdesai) [reply backends/arm/arm_backend.py:167]: I meant AoT generates something which Runtime consumes and we verify the results - i.e. e2e test. So we don't have to worry about making sure AoT and runtime are aligned.

**Pull request closed without merging**